compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/Utils.cpp

```diff
@@ -46,19 +45,49 @@ std::tuple<Value, Value> lookupDeviceAndQueueAffinityFor(Operation *op,
   return std::make_tuple(resolveOp.getResult(0), resolveOp.getResult(1));
 }
 
-std::tuple<Value, Value> lookupDeviceAndQueueAffinityFor(
-    Operation *op, IREE::HAL::MemoryTypeBitfield memoryTypes,
-    IREE::HAL::BufferUsageBitfield bufferUsage, OpBuilder &builder) {
-  // Emit a select op to let the runtime decide which device/queue affinity to
-  // use if required.
+static std::tuple<SmallVector<Value>, SmallVector<Value>>
+lookupDevicesAndQueueAffintiesFor(Operation *op, OpBuilder &builder) {
   auto affinityAttr = IREE::Stream::AffinityAttr::lookupOrDefault(op);
+  SmallVector<Value> devices;
+  SmallVector<Value> queueAffinities;
   if (auto optimalAttr = dyn_cast<IREE::HAL::DeviceOptimalAttr>(affinityAttr)) {
-    auto selectOp = builder.create<IREE::HAL::AllocatorSelectAttrOp>(
-        op->getLoc(), optimalAttr, memoryTypes, bufferUsage);
-    return {selectOp.getResult(0), selectOp.getResult(1)};
+    for (auto device : optimalAttr.getAffinities()) {
+      std::tuple<Value, Value> deviceAndQueueAffinity =
+          lookupDeviceAndQueueAffinityFor(op->getLoc(), device, builder);
+      devices.push_back(std::get<0>(deviceAndQueueAffinity));
+      queueAffinities.push_back(std::get<1>(deviceAndQueueAffinity));
```

Comment:
C++ structured bindings make tuples much better: https://en.cppreference.com/w/cpp/language/structured_binding.html

```suggestion
      auto [device, queueAffinity] =
          lookupDeviceAndQueueAffinityFor(op->getLoc(), device, builder);
      devices.push_back(device);
      queueAffinities.push_back(queueAffinity);
```